### PR TITLE
Add Release Instructions + Release Github Action

### PR DESCRIPTION
**Pull request by solbrigm**
## Summary (What and Why)

Adding CONTRIBUTING.md with contributing guidelines and release instructions.

Adding a github action to release when a release is published.

## Test Plan

When it comes time to release, there will likely be some debugging involved :( 

## Release Plan

NA

## GitHub Issue:

NA

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,6 @@ docs/build/
 
 # Hatch generated files
 dist/
+
+# venvs
+*venv/
```
Comment (chrisjordansquire) [inline .gitignore:18]: Is there a reason the venvs would be in the source directory? (I think I've always studiously avoided having venvs or builds in a source directory, to the extent I could.)

**Review by chrisjordansquire — COMMENTED** [review]
A few small suggestions. I'll defer to @brian-tecton-ai to make a call on the release yml, since I'm guessing that relates to an already-existing release guideline.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# Contributing Guidelines
+
+## Reporting Bugs
+
+If you encounter a bug, please help us by submitting a detailed bug report. Include information such as:
+
+* Steps to reproduce the bug
+* Expected behavior
+* Actual behavior
+* Environment details (operating system, Python version, etc.)
```
[the remark this answers is not in the document]
Comment (solbrigm) [reply CONTRIBUTING.md:10]: Good point, will do. (Full disclosure: this advice came from asking ChatGPT to generate contributing guidelines for me)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+along with any relevant use cases or examples.
+
+## Pull Requests
+
+* Fork the repository and create a new branch for your contribution.
+* Ensure your code adheres to the project's coding style and conventions.
+* Write clear, concise commit messages.
+* Include tests for any new functionality or changes.
+* Update documentation as needed.
+* Submit a pull request, explaining the purpose of your changes.
```
[the remark this answers is not in the document]
Comment (solbrigm) [reply CONTRIBUTING.md:24]: They're enforced by the pre-commit hook, so don't necessarily seem worth spelling out in too much detail, but let me add at least a little of detail and make sure that installing the pre-commit hooks is in the dev setup.

File: .gitignore (hunk 1 above)
Comment (solbrigm) [reply .gitignore:18]: TBH I agree; I keep my virtual envs in a centralized folder. However, a lot of IDEs will create the virtual env in the project folder by default. It's also much easier to give instructions for setting up an environment if assume the venv is created locally. Given that, it seems worth excluding the `*venv` pattern, thus supporting people to create venv within the project, unless there is a good reason to preserve it.

Another reason that isn't relevant in this case, is allows for self-contained `make` targets for the project, where that is popular (`make test`, `make clean`, `make build`, etc).

**Review by chrisjordansquire — APPROVED** [review]

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+To set up the project for development:
+
+* (suggested) Create a virtual env: `python -m venv test_venv; . ./test_venv/bin/activate`
+* Install the pre-commit hooks:
+```bash
+pip install pre-commit
+pre-commit install
+```
+* Install dependencies: `pip install .[test]`
```
Comment (brian-tecton-ai) [inline CONTRIBUTING.md:22]: shoudl this be `pip install -e .[test]`?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      - uses: actions/checkout@v3
+      - name: Install Hatch
+        run: pipx install hatch
+      - name: Run tests
+        run: hatch run test:test
+      - name: Build dist
+        run: hatch build
+      - name: Publish on PyPI
+        if: github.ref == 'refs/heads/main'
+        run: hatch publish
```
Comment (brian-tecton-ai) [inline .github/workflows/release.yml:22]: maybe we can pull some inspiration by how this works:
https://github.com/pypa/hatch/blob/8e0db522be51f47e189b9357c5d11b03787ca6a7/.github/workflows/build-hatch.yml#L440-L483

Some things that I like here that we should take inspiration from:
* breaks down the steps into a few different jobs (separation of build, test, publish). This makes for better retries / observability into a failed workflow.
* OIDC auth for publishing to PyPI. More secure than using secrets. https://github.com/marketplace/actions/pypi-publish#trusted-publishing
* GitHub action is responsible for creating the release itself https://github.com/softprops/action-gh-release (so our wheel artifact is actually part of the GitHub release)

Comment (solbrigm) [reply .github/workflows/release.yml]: Updated; I've broken the pipeline into build vs publish, but keeping the release as manual so we don't have to automate the release notes.

**Pull request merged**